File: packages/helm-runtime-sdk/node/src/proxy.js
// The same-origin proxy a studio mounts at /helm/, so its page never holds a
// token. It behaves exactly as the Go and Python runtime SDKs' proxies do
// (docs/decisions.md M6 Q10; test/conformance holds all three to it):
//
//   /helm/api/v1/<studio-api path>  forwarded to HELM_API with HELM_TOKEN added;
//                                   only the studio API and the theme stream
//   /helm/sdk/v1/<file>             GET and HEAD, to HELM_SDK_BASE, no token
//   /helm/accent.css                the studio's hue for each theme
//
// The page's own Authorization, cookies, Origin and Referer are never
// forwarded, and the daemon's Set-Cookie never comes back.
//
//   import http from "node:http";
//   import { createProxy } from "@helmstudio/runtime/proxy";
//   const helmProxy = createProxy();
//   http.createServer(async (req, res) => {
//     if (await helmProxy(req, res)) return;
//     // …the studio's own routes
//   }).listen(port, "127.0.0.1");
//
// Uses fetch and Node's request and response objects; not for the browser.

export const PREFIX = "/helm/";

export const STUDIO_SEGMENTS = new Set([
  "me", "events", "kv", "sessions", "records", "assets", "assets:adopt", "gallery", "handoff", "inbox", "jobs", "theme",
]);

const REQUEST_HEADERS = ["Accept", "Content-Type", "Range", "If-Range", "If-Match", "If-None-Match", "Last-Event-ID"];
const RESPONSE_HEADERS = ["Content-Type", "Content-Length", "Content-Range", "Accept-Ranges", "ETag", "Last-Modified", "Cache-Control", "Content-Disposition", "Allow"];
const METHODS = new Set(["GET", "HEAD", "POST", "PUT", "PATCH", "DELETE"]);
const HEX = /^#[0-9a-fA-F]{6}$/;

function luminance(hex) {
  const ch = [0, 1, 2].map((i) => {
    const c = parseInt(hex.slice(1 + 2 * i, 3 + 2 * i), 16) / 255;
    return c <= 0.04045 ? c / 12.92 : Math.pow((c + 0.055) / 1.055, 2.4);
  });
  return 0.2126 * ch[0] + 0.7152 * ch[1] + 0.0722 * ch[2];
}

function on(hex) {
  const l = luminance(hex);
  const dark = (l + 0.05) / (luminance("#1a1400") + 0.05);
  const white = 1.05 / (l + 0.05);
  return dark >= white ? "#1a1400" : "#ffffff";
}

export function accentCSS(dark, light) {
  if (!dark || !light || !HEX.test(dark) || !HEX.test(light)) return null;
  dark = dark.toLowerCase();
  light = light.toLowerCase();
  const block = (h) => `--helm-studio-accent: ${h}; --helm-on-studio-accent: ${on(h)};`;
  return (
    `:root { ${block(dark)} }\n` +
    `@media (prefers-color-scheme: light) { :root:not([data-theme]) { ${block(light)} } }\n` +
    `:root[data-theme="light"] { ${block(light)} }\n`
  );
}

export function safePath(p) {
  const lower = p.toLowerCase();
  if (lower.includes("%2f") || lower.includes("%5c") || lower.includes("%2e") || p.includes("\\")) return false;
  const segs = p.replace(/^\//, "").split("/");
  return segs.every((s, i) => s !== "." && s !== ".." && !(s === "" && i !== segs.length - 1));
}

function send(res, status, headers, body) {
  res.writeHead(status, headers);
  res.end(body);
}

function notFound(res) {
  send(res, 404, { "Content-Type": "application/json", "X-Content-Type-Options": "nosniff" },
    JSON.stringify({ error: "not_found", message: "the helm proxy forwards only the studio API, the theme stream, the SDK files and accent.css" }) + "\n");
}

export function createProxy({ env = process.env, fetch: fetchImpl = globalThis.fetch } = {}) {
  const api = (env.HELM_API || "").replace(/\/+$/, "");
  const token = env.HELM_TOKEN || "";
  const sdkBase = (env.HELM_SDK_BASE || "").replace(/\/+$/, "");
  const css = accentCSS(env.HELM_ACCENT_DARK, env.HELM_ACCENT_LIGHT);

  async function forward(req, res, target, withToken, query) {
    if (!METHODS.has(req.method)) {
      send(res, 405, { "Content-Type": "application/json" }, JSON.stringify({ error: "method_not_allowed", message: `${req.method} is not forwarded` }) + "\n");
      return;
    }
    const headers = {};
    for (const h of REQUEST_HEADERS) {
      const v = req.headers[h.toLowerCase()];
      if (v !== undefined) headers[h] = v;
    }
    if (withToken && token) headers.Authorization = `Bearer ${token}`;
    const init = { method: req.method, headers, redirect: "manual" };
    if (req.method !== "GET" && req.method !== "HEAD") {
      init.body = req;
      init.duplex = "half";
    }
    const abort = new AbortController();
    init.signal = abort.signal;
    res.on("close", () => abort.abort());
    let upstream;
    try {
      upstream = await fetchImpl(target + (query ? `?${query}` : ""), init);
    } catch {
      // Never the error text: it can carry the upstream URL.
      if (!res.headersSent) {
        send(res, 503, { "Content-Type": "application/json" }, JSON.stringify({ error: "unavailable", message: "helmstudio did not answer" }) + "\n");
      }
      return;
    }
    const out = { "X-Content-Type-Options": "nosniff" };
    for (const h of RESPONSE_HEADERS) {
      const v = upstream.headers.get(h);
      if (v !== null) out[h] = v;
    }
    res.writeHead(upstream.status, out);
    if (req.method === "HEAD" || !upstream.body) {
      res.end();
      return;
    }
    try {
      for await (const chunk of upstream.body) {
        if (!res.write(chunk)) await new Promise((r) => res.once("drain", r));
      }
    } catch {
      // the page went away, or the daemon did
    }
    res.end();
  }

  return async function handle(req, res) {
    const [path, query = ""] = req.url.split(/\?(.*)/s);
    if (!path.startsWith(PREFIX)) return false;
    const rest = path.slice(PREFIX.length - 1);
    if (!safePath(rest)) {
      notFound(res);
      return true;
    }
    if (rest === "/accent.css") {
      if (!css || (req.method !== "GET" && req.method !== "HEAD")) notFound(res);
      else send(res, 200, { "Content-Type": "text/css; charset=utf-8", "Cache-Control": "no-cache", "X-Content-Type-Options": "nosniff" }, req.method === "HEAD" ? undefined : css);
      return true;
    }
    if (rest.startsWith("/api/v1/")) {
      const sub = rest.slice("/api/v1/".length);
      if (!api || !STUDIO_SEGMENTS.has(sub.split("/")[0])) notFound(res);
      else await forward(req, res, `${api}/${sub}`, true, query);
      return true;
    }
    if (rest.startsWith("/sdk/v1/")) {
      if (!sdkBase || (req.method !== "GET" && req.method !== "HEAD")) notFound(res);
      else await forward(req, res, `${sdkBase}/${rest.slice("/sdk/v1/".length)}`, false, query);
      return true;
    }
    notFound(res);
    return true;
  };
}

File: packages/helm-runtime-sdk/go/proxy.go
package helm

import (
	"encoding/json"
	"io"
	"math"
	"net/http"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// Environment variables the daemon injects at spawn for a studio's page
// (docs/design/07-platform-services.md §3, docs/decisions.md M6 Q7, Q8, Q14).
const (
	EnvTheme       = "HELM_THEME"
	EnvAccentDark  = "HELM_ACCENT_DARK"
	EnvAccentLight = "HELM_ACCENT_LIGHT"
	EnvSDKBase     = "HELM_SDK_BASE"
)

// ProxyPrefix is where a studio mounts the proxy on its own server.
const ProxyPrefix = "/helm/"

// ProxyConfig is what the proxy forwards to. Empty API or SDKBase means that
// half answers 404, as it does standalone; an empty Token forwards without
// one (a studio with no capabilities has none).
type ProxyConfig struct {
	API         string // HELM_API, e.g. http://127.0.0.1:8700/api/v1
	Token       string // HELM_TOKEN
	SDKBase     string // HELM_SDK_BASE, e.g. http://127.0.0.1:8700/sdk/v1
	AccentDark  string // HELM_ACCENT_DARK
	AccentLight string // HELM_ACCENT_LIGHT
	// Client sends the forwarded requests; nil is a client with no timeout,
	// since event streams stay open.
	Client *http.Client
}

// ProxyFromEnv reads ProxyConfig from the variables the daemon and helm dev
// inject.
func ProxyFromEnv() ProxyConfig {
	return ProxyConfig{API: os.Getenv(EnvAPI), Token: os.Getenv(EnvToken), SDKBase: os.Getenv(EnvSDKBase),
		AccentDark: os.Getenv(EnvAccentDark), AccentLight: os.Getenv(EnvAccentLight)}
}

// Proxy is the same-origin proxy a studio mounts at /helm/ on its own server,
// so its page can use the platform API without ever holding a token
// (docs/decisions.md M6 Q10):
//
//	mux.Handle(helm.ProxyPrefix, helm.Proxy(helm.ProxyFromEnv()))
//
// It answers exactly three things:
//   - /helm/api/v1/<studio-api path> — forwarded to HELM_API with the studio's
//     token added. Only the studio API and the theme stream: a launcher path
//     (install, launch, stop, reclaim) is 404 and never reaches the daemon.
//   - /helm/sdk/v1/<file> — GET and HEAD, forwarded to HELM_SDK_BASE without a
//     token.
//   - /helm/accent.css — the studio's hue as --helm-studio-accent and
//     --helm-on-studio-accent for each theme (03 §2c).
//
// The page's own Authorization, cookies, Origin and Referer are never
// forwarded, and the daemon's Set-Cookie is never returned. The Python and
// Node runtime SDKs ship the same proxy; test/conformance holds all three to
// the same behaviour.
func Proxy(cfg ProxyConfig) http.Handler {
	if cfg.Client == nil {
		cfg.Client = &http.Client{}
	}
	return &proxy{cfg: cfg}
}

type proxy struct{ cfg ProxyConfig }

// ProxyStudioSegments are the first path segments under /api/v1/ the proxy
// forwards: every studio-api operation's, and the tokenless theme stream's.
var ProxyStudioSegments = map[string]bool{
	"me": true, "events": true, "kv": true, "sessions": true, "records": true,
	"assets": true, "assets:adopt": true, "gallery": true, "handoff": true,
	"inbox": true, "jobs": true, "theme": true,
	"timeline": true, "timeline:append": true,
}

// Request headers a page may send through; everything else is dropped.
var proxyRequestHeaders = []string{"Accept", "Content-Type", "Range", "If-Range", "If-Match", "If-None-Match", "Last-Event-ID"}

// Response headers returned to the page.
var proxyResponseHeaders = []string{"Content-Type", "Content-Length", "Content-Range", "Accept-Ranges", "Etag", "Last-Modified", "Cache-Control", "Content-Disposition", "Allow"}

func (p *proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	rest, ok := strings.CutPrefix(r.URL.EscapedPath(), strings.TrimSuffix(ProxyPrefix, "/"))
	if !ok || !safePath(rest) {
		proxyNotFound(w, r)
		return
	}
	switch {
	case rest == "/accent.css":
		p.accent(w, r)
	case strings.HasPrefix(rest, "/api/v1/"):
		sub := strings.TrimPrefix(rest, "/api/v1/")
		first, _, _ := strings.Cut(sub, "/")
		if p.cfg.API == "" || !ProxyStudioSegments[first] {
			proxyNotFound(w, r)
			return
		}
		p.forward(w, r, strings.TrimRight(p.cfg.API, "/")+"/"+sub, p.cfg.Token)
	case strings.HasPrefix(rest, "/sdk/v1/"):
		if p.cfg.SDKBase == "" || (r.Method != http.MethodGet && r.Method != http.MethodHead) {
			proxyNotFound(w, r)
			return
		}
		p.forward(w, r, strings.TrimRight(p.cfg.SDKBase, "/")+"/"+strings.TrimPrefix(rest, "/sdk/v1/"), "")
	default:
		proxyNotFound(w, r)
	}
}

// safePath refuses a path that could leave the prefix it was checked under
// once the upstream decodes it: dot segments, empty segments, and encoded
// slashes, backslashes or dots.
func safePath(p string) bool {
	lower := strings.ToLower(p)
	if strings.Contains(lower, "%2f") || strings.Contains(lower, "%5c") || strings.Contains(lower, "%2e") || strings.Contains(p, "\\") {
		return false
	}
	segs := strings.Split(strings.TrimPrefix(p, "/"), "/")
	for i, s := range segs {
		if s == "." || s == ".." || (s == "" && i != len(segs)-1) {
			return false
		}
	}
	return true
}

func (p *proxy) forward(w http.ResponseWriter, r *http.Request, target, token string) {
	switch r.Method {
	case http.MethodGet, http.MethodHead, http.MethodPost, http.MethodPut, http.MethodPatch, http.MethodDelete:
	default:
		proxyError(w, http.StatusMethodNotAllowed, "method_not_allowed", r.Method+" is not forwarded")
		return
	}
	if r.URL.RawQuery != "" {
		target += "?" + r.URL.RawQuery
	}
	var body io.Reader
	if r.Body != nil && r.Method != http.MethodGet && r.Method != http.MethodHead {
		body = r.Body
	}
	out, err := http.NewRequestWithContext(r.Context(), r.Method, target, body)
	if err != nil {
		proxyError(w, http.StatusBadGateway, "proxy_error", "the request could not be forwarded")
		return
	}
	if body != nil {
		out.ContentLength = r.ContentLength
	}
	for _, h := range proxyRequestHeaders {
		if v := r.Header.Values(h); len(v) > 0 {
			out.Header[h] = append([]string(nil), v...)
		}
	}
	if token != "" {
		out.Header.Set("Authorization", "Bearer "+token)
	}
	res, err := p.cfg.Client.Do(out)
	if err != nil {
		if r.Context().Err() == nil {
			// Never the error text: it can carry the upstream URL.
			proxyError(w, http.StatusServiceUnavailable, "unavailable", "helmstudio did not answer")
		}
		return
	}
	defer res.Body.Close()
	for _, h := range proxyResponseHeaders {
		if v := res.Header.Values(h); len(v) > 0 {
			w.Header()[h] = append([]string(nil), v...)
		}
	}
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.WriteHeader(res.StatusCode)
	if r.Method == http.MethodHead {
		return
	}
	rc := http.NewResponseController(w)
	buf := make([]byte, 32*1024)
	for {
		n, err := res.Body.Read(buf)
		if n > 0 {
			if _, werr := w.Write(buf[:n]); werr != nil {
				return
			}
			_ = rc.Flush() // event streams arrive as they are written
		}
		if err != nil {
			return
		}
	}
}

var hexColour = regexp.MustCompile(`^#[0-9a-fA-F]{6}$`)

func (p *proxy) accent(w http.ResponseWriter, r *http.Request) {
	css, ok := AccentCSS(p.cfg.AccentDark, p.cfg.AccentLight)
	if !ok || (r.Method != http.MethodGet && r.Method != http.MethodHead) {
		proxyNotFound(w, r)
		return
	}
	w.Header().Set("Content-Type", "text/css; charset=utf-8")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	if r.Method == http.MethodGet {
		_, _ = io.WriteString(w, css)
	}
}

// AccentCSS is /helm/accent.css for a hue pair: --helm-studio-accent and
// --helm-on-studio-accent per theme, following data-theme the way
// helm-tokens.css does (03 §2c). ok is false unless both are "#rrggbb".
func AccentCSS(dark, light string) (string, bool) {
	if !hexColour.MatchString(dark) || !hexColour.MatchString(light) {
		return "", false
	}
	dark, light = strings.ToLower(dark), strings.ToLower(light)
	block := func(hex string) string {
		return "--helm-studio-accent: " + hex + "; --helm-on-studio-accent: " + onColour(hex) + ";"
	}
	return ":root { " + block(dark) + " }\n" +
		"@media (prefers-color-scheme: light) { :root:not([data-theme]) { " + block(light) + " } }\n" +
		":root[data-theme=\"light\"] { " + block(light) + " }\n", true
}

// onColour is #1a1400 or #ffffff, whichever contrasts more with hex.
func onColour(hex string) string {
	l := luminance(hex)
	dark := (l + 0.05) / (luminance("#1a1400") + 0.05)
	white := (1.05) / (l + 0.05)
	if dark >= white {
		return "#1a1400"
	}
	return "#ffffff"
}

func luminance(hex string) float64 {
	var rgb [3]float64
	for i := range rgb {
		v, _ := strconv.ParseUint(hex[1+2*i:3+2*i], 16, 8)
		c := float64(v) / 255
		if c <= 0.04045 {
			rgb[i] = c / 12.92
		} else {
			rgb[i] = math.Pow((c+0.055)/1.055, 2.4)
		}
	}
	return 0.2126*rgb[0] + 0.7152*rgb[1] + 0.0722*rgb[2]
}

func proxyNotFound(w http.ResponseWriter, r *http.Request) {
	proxyError(w, http.StatusNotFound, "not_found", "the helm proxy forwards only the studio API, the theme stream, the SDK files and accent.css")
}

func proxyError(w http.ResponseWriter, status int, code, msg string) {
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(map[string]string{"error": code, "message": msg})
}
